File: Week7/alotofhee/오픈채팅방.java
import java.util.*;
class Solution {
    public String[] solution(String[] record) {
        String[] answer = {};
        HashMap<String, String> info = new HashMap<>();
        StringBuilder sb = new StringBuilder();
        int count = record.length;
      
        for (int i = 0; i < record.length; i++) {
            String func = record[i].split(" ")[0];
            String id = record[i].split(" ")[1];
            String nickname = "";          
            if (!func.equals("Leave")) {
                nickname = record[i].split(" ")[2];
            }            
            switch(func) {
                case "Enter":
                    sb.append(id).append("님이 들어왔습니다.-");
                    info.put(id, nickname);
                    break;
                case "Leave":
                    sb.append(id).append("님이 나갔습니다.-");
                    break;
                case "Change":
                    info.put(id, nickname);
                    count--;
                    break;
            }
        }
        
        String str = sb.toString();
        answer = new String[count];
        answer = str.split("-");
        for(int i = 0; i < answer.length; i++){
            String key = answer[i].substring(0, answer[i].indexOf("님"));
            answer[i] = answer[i].replace(key, info.get(key));
        }                      
        return answer;
    }
}

File: Week7/dlduddk/오픈채팅방.py
"""
문제 접근
    나갔다가 다시 들어올 때, 또는 이름 변경 -> 들어오고 나갈 때 해당 유저가 사용했던 닉네임 모두 바뀜
해결 전략
    딕셔너리에 key : 유저 아이디 value : 최종 닉네임 저장
"""
def solution(record):
    answer = []
    user = {}

    for i in record: 
        a = i.split(' ')
        if a[0] == 'Enter' or a[0] == 'Change':
            user[a[1]] = a[2]

    
    for i in record:
        a = i.split(' ')
        if a[0] == 'Enter': 
            answer.append(user[a[1]]+"님이 들어왔습니다.")
        elif a[0] == 'Leave': 
            answer.append(user[a[1]]+"님이 나갔습니다.")
        
    return answer